src/extensionApi.ts

```diff
@@ -262,16 +262,28 @@ export class CommandHandler {
             context = this.explorer.getServerStateById(rsp.id, serverId);
         }
 
-        return this.stopServer(false, context)
-            .then(() => {
+        const client: RSPClient = this.explorer.getClientByRSP(context.rsp);
+        if (!client) {
+            return Promise.reject('Failed to contact the RSP server.');
+        }
+
+        const listener = (state: Protocol.ServerState) => {
+            if (state
+                && state.server
+                && state.server.id === context.server.id
+                && state.state === ServerState.STOPPED) {
+                client.getIncomingHandler().removeOnServerStateChanged(listener);
                 if (mode === 'debug') {
```

Comment:
maybe a switch would be more expressive than the sequence of if/else if/else

Reply:
furthermore I'd also highly encourage defining the mode as a constant. I see 'debug' and 'run' being used and repeated in several place. Once we'd change a mode, we'd have correct all these occurrences